feat: page reports v0.0.1

lricoy opened the pull request:
## Problem

Web Analytics users need a dedicated place to view detailed metrics for specific pages on their website. Currently, users must navigate multiple tabs and tables in the main dashboard to find page-specific data, making it difficult to get a comprehensive view of how individual pages perform.
This feature addresses the need for a centralized view of page-specific analytics, (hopefully) making it easier for users to analyze and optimize individual pages. It will be the groundwork for future integration with other PostHog products, such as heatmaps, session replay, and SEO data.

## Changes

This PR introduces Page Reports v0.0.1, which provides a dedicated view for analyzing individual page performance. It is a simplified version of the Web Analytics dashboard without the tabs.

I've decided for it to have its own Kea Logic to avoid side effects with the main dashboard as much as possible.

<img width="1726" alt="image" src="https://github.com/user-attachments/assets/22fa5be9-bbea-4fab-a8a2-90874ebe026a" />

## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

- Manually tested the page selection functionality with various URLs
- Verified that all metrics load correctly for selected pages

## Future iterations will include:
- Charts/Tables as in the main dashboard
- StatsOverviews
- Heatmaps Preview
- Compare with previous period + path cleaning
- Top events on page + correlation

lricoy committed:
chore: boilperplate for new tab

lricoy committed:
simple tables working

lricoy committed:
fix url refresh not setting correct tab

lricoy committed:
added remaining tables

lricoy committed:
added demographics

lricoy committed:
add a combined graph chart

lricoy committed:
finishing scafolding

lricoy committed:
make it look a bit more like the rest of wa

lricoy committed:
fix ts errors

lricoy committed:
some cleanup

lricoy committed:
wip searchbar for url

lricoy committed:
v1 of search bar working

lricoy committed:
working search

lricoy committed:
added FF logic and alpha tag

lricoy committed:
best effort to make it work with path cleaning

lricoy committed:
somewhat working strip query parameters

lricoy committed:
polished kea logic

lricoy committed:
wip: added trends visualizations as in web analytics

lricoy committed:
wip logic

lricoy committed:
working refactor

lricoy committed:
removed dateFilter link to webAnalytics

lricoy committed:
use contains for stripQueryParameters

lricoy committed:
tiding up to use less code

lricoy committed:
tiding up

lricoy committed:
Remove code for first PR

lricoy committed:
Update pageReportsLogic to use regex for URL matching when stripQueryParams is true

lricoy committed:
allow filtering stats table with url

lricoy committed:
fixed filter not working on stats tables

lricoy committed:
Convert strip query parameters into filter button

lricoy committed:
make page report filters sticky

rafaeelaudibert commented on frontend/src/scenes/web-analytics/WebAnalyticsHealthCheck.tsx:
```diff
@@ -49,7 +49,10 @@ export const WebAnalyticsHealthCheck = (): JSX.Element | null => {
     }
 
-    if (productTab === ProductTab.WEB_VITALS) {
+    // For Page reports tab, use the same health check as the Analytics tab
```
This change seems moot, we could have kept the previous code

lricoy replied:
Agreed. Will decursorfy it.

rafaeelaudibert reviewed (commented):
Initial review!

Looking good for now, but I have some questions:

- why not reuse WA filters?
- why change from tiles to queries? I dont *love* our tile system, but if we wanna change it then we should 100% change it


Suggestions:
- This is already pretty good! Unless you feel like something's janky then I suggest you make it an early access feature and label it as alpha, this will get you some early testers. You can then create a survey to collect feedback
- The heatmaps UI will be useless after you add this, you should check with Paul what that means to Heatmaps as a product: do we own it now? Will it go away from the sidebar?

lricoy commented:
@rafaeelaudibert thanks for the initial review! 

Awesome questions! 

> why not reuse WA filters?

Allowing them was causing too much trouble because it is super easy to filter by path. The tables allow it; we can use `$current_url`, `pathname`, `host`, etc. The idea here would be to choose a single URL to look at. I don't know if we'll be able to use the same ones here because that would still mean we would choose two controls (host + path) to make it work, but I am open to suggestions there. 

> why change from tiles to queries? I dont love our tile system, but if we wanna change it then > we should 100% change it

Yeah, I am also not happy with how that turned out. It started because I wanted more flexibility to design the page with the other components (heatmaps, replays, events, etc.), and using simple react components was easier for me. But after removing most of the code here to open the first PR, there are only tile-related components, so I am considering converting them to tiles-state as in WebAnalyticsDashboard. 

Suggestions:

> This is already pretty good! Unless you feel like something's janky then I suggest you make it an early access feature and label it as alpha, this will get you some early testers. You can then create a survey to collect feedback

Thanks! I will wait for Andy, Paul, and Lucas' feedback (i.e, tomorrow) before choosing what to do, but those are great options.

> The heatmaps UI will be useless after you add this, you should check with Paul what that means to Heatmaps as a product: do we own it now? Will it go away from the sidebar?

I was thinking about only adding a summarized preview that would then provide a link or open a modal (like we do with session replay) so we can tighten our link with one more product, but that was also one of my questions to Paul tomorrow.

lricoy committed:
Merge branch 'master' into feat/page-reports-v1.1

lricoy committed:
review suggestions

pauldambra committed:
fix: celery scheduling doubling (#29853)

lricoy committed:
keep dashboard component simple

lricoy commented on frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx:
```diff
@@ -2287,10 +2331,19 @@ export const webAnalyticsLogic = kea<webAnalyticsLogicType>([
+            const effectiveProductTab =
+                !pageReportsEnabled && productTab === ProductTab.PAGE_REPORTS ? ProductTab.ANALYTICS : productTab
+
+            if (effectiveProductTab !== ProductTab.ANALYTICS) {
```
Good question. Looking at the code, I don't think that the `product_tab` query parameter is required anymore. Maybe that was prior to being part of the url path like `/web/vitals`? Testing locally it seems safe to remove it.

rafaeelaudibert replied:
Oh yeah, I guess you're right. It was from a time where we didn't have a custom URL for web analytics. It's time to go :)

lricoy committed:
more cleanup after review

lricoy committed:
more cleanup after review

darkopia committed:
fix: update copy to pluralize correctly (#29887)

Co-authored-by: github-actions <41898282+github-actions[bot]@users.noreply.github.com>

github-actions[bot] committed:
Update UI snapshots for `webkit` (2)

github-actions[bot] committed:
Update UI snapshots for `webkit` (2)

github-actions[bot] committed:
Update UI snapshots for `webkit` (2)

github-actions[bot] committed:
Update UI snapshots for `webkit` (2)

github-actions[bot] committed:
Update UI snapshots for `webkit` (2)

lricoy committed:
makes greptile and TS happy

github-actions[bot] committed:
Update UI snapshots for `webkit` (2)

github-actions[bot] committed:
Update UI snapshots for `webkit` (2)

lricoy committed:
Merge branch 'master' into feat/page-reports-v1.1

robbie-c commented on frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx:
```diff
@@ -112,6 +129,22 @@ const loadPriorityMap: Record<TileId, number> = {
     [TileId.WEB_VITALS_PATH_BREAKDOWN]: 12,
+
+    // Page Report Tiles
+    [TileId.PAGE_REPORTS_COMBINED_METRICS_CHART]: 13,
```
You could start these at 1 if you like, it doesn't really matter as we shouldn't have an active query for both PageReports and the regulaer dashboard at the same time

robbie-c commented on frontend/src/scenes/web-analytics/pageReportsLogic.ts:
```diff
@@ -0,0 +1,343 @@
+            (selectors) => [selectors.pagesLoading, selectors.isInitialLoad],
+            (pagesLoading: boolean, isInitialLoad: boolean) => pagesLoading || isInitialLoad,
+        ],
+        queries: [
```
Hmm I wonder if we should refactor this.

Instead of
* webAnalyticsLogic has a selector that creates all the queries and puts them in an array for the UI component to consume
* pageReportsLogic has a selector that takes the queries out of the array to put on an object for page reports UI consume

We could do
* webAnalyticsLogic has a selector that creates all the queries an puts them on an object
* webAnalyticsLogic has a selector that takes them from the object and puts in an array for the UI component to consume
* pageReportsLogic has a selector that takes the queries out of the *object* to put on an object for page reports to consume

That way your queries wouldn't need to be optional in `SimpleTile`

lricoy replied:
I agree with that path, but this PR got ~2.5 times bigger after doing it because of the many tileId + tabId combinations required. 

In the end, I did convert the logic here to use `<Tiles />` as well, so I suggest putting that refactor in another PR.

WDYT?

robbie-c commented on frontend/src/scenes/scenes.ts:
```diff
@@ -541,6 +547,7 @@ export const routes: Record<string, [Scene | string, string]> = {
     [urls.savedInsights()]: [Scene.SavedInsights, 'savedInsights'],
     [urls.webAnalytics()]: [Scene.WebAnalytics, 'webAnalytics'],
     [urls.webAnalyticsWebVitals()]: [Scene.WebAnalytics, 'webAnalyticsWebVitals'],
+    [urls.webAnalyticsPageReports()]: [Scene.WebAnalytics, 'webAnalyticsPageReports'],
```
I'm getting that this doesn't exist, I think atm you need to add it to both `products` and `urls`

lricoy replied:
I think I had the same issue as you did when running after merging with master. 

Urls now need to be placed on: `product folder manifests under products/*/manifest.tsx`.
after #29728

robbie-c commented:
<img width="1581" alt="Screenshot 2025-03-14 at 15 38 08" src="https://github.com/user-attachments/assets/c7ec9d48-4c9c-490c-a594-1f97084a40b3" />

I get this error when changing between having a URL set and not

robbie-c commented:
I do agree that I'd prefer to only have 1 of tiles or components, I'm not a huge fan of having two different patterns. For the sake of keeping the PR small I'd probably go with a tiles-based approach, but if you want to move everything to a component-based approach then you could.

lricoy committed:
start pageReports'tiles loadPriorityMap from 1

lricoy committed:
prefer general kea pratices

lricoy committed:
add page reports to manifest.tsx

lricoy committed:
fix react hooks render order

lricoy committed:
convert pageReports to use tiles

lricoy committed:
make WebSectionTile more reusable

lricoy committed:
make WebSectionTile more reusable

rafaeelaudibert commented on frontend/src/scenes/web-analytics/PageReports.tsx:
```diff
@@ -0,0 +1,93 @@
+                        onChange={(val: string[]) => setPageUrl(val.length > 0 ? val[0] : null)}
+                        options={options}
+                        onInputChange={(val: string) => setPageUrlSearchTerm(val)}
+                        data-attr="page-url-search"
```
```suggestion
                        data-attr="page-reports-url-search"
```

rafaeelaudibert commented on frontend/src/scenes/web-analytics/PageReports.tsx:
```diff
@@ -0,0 +1,93 @@
+                    <DateFilter
+                        dateFrom={dateFilter.dateFrom}
+                        dateTo={dateFilter.dateTo}
+                        onChange={(fromDate, toDate) => setDates(fromDate, toDate)}
```
```suggestion
                        onChange={setDates}
```

rafaeelaudibert commented on frontend/src/scenes/web-analytics/PageReports.tsx:
```diff
@@ -0,0 +1,93 @@
+        return (
+            <div className="space-y-2 mt-2">
+                <ProductIntroduction
+                    productName="PAGE REPORTS"
```
I don't see this as uppercase anywhere else, is this correct?
![image](https://github.com/user-attachments/assets/72d35044-4f08-48a5-a8be-45241cd61a5b)

lricoy committed:
use tile's layout props to style

rafaeelaudibert reviewed (commented):
3 nitpicks.

I agree with Robbie's proposed structure [here](https://github.com/PostHog/posthog/pull/29902/files#r1995770493)

lricoy committed:
apply review suggestions and some ui improvements

lricoy committed:
feat: disable web stats table clicks on page reports

lricoy committed:
chore: improve the ux writing

lricoy committed:
chore: simplify SectionItemTile

lricoy committed:
Discard changes to frontend/src/scenes/urls.ts

lricoy commented:
> I get this error when changing between having a URL set and not.
> 
> I couldn't figure out a super obvious place where it was coming from though!
> 
> One difference between your code and other kea code I've seen is that you sometimes do this
> 
> ```
> const values = useValues(pageReportsLogic)
> ```
> 
> instead of
> 
> ```
> const { dateFilter } = useValues(webAnalyticsLogic)
> ```
> 
> (you wrote both of those :D). Anyway I wonder if accessing `values` does some magic behind the scenes, that would be the first place I would look, just change your use of `useValues` to be more like the second example.

Nice. I had not seen that one before, but after checking the merged code, it was happening every time I removed the `selectedUrl`. 

It turns out there was no magic there for Kea, just automated refactors that kept it as `values` instead of restructuring. 

If I got it right, the problem was that it was relying on `pageUrl` being present for the hooks in the `SimpleTiles`, but that would not be the case after it was removed, so it was missing the checks to not re-render those.


> I do agree that I'd prefer to only have 1 of tiles or components, I'm not a huge fan of having two different patterns. For the sake of keeping the PR small I'd probably go with a tiles-based approach, but if you want to move everything to a component-based approach then you could.

Agreed! I converted it to use `<Tiles />` instead, but would still rather keep the state in another logic as mentioned on the other comment to avoid side effects on the `webAnalyticsLogic` for now.

lricoy committed:
Merge branch 'master' into feat/page-reports-v1.1

lricoy committed:
Merge branch 'master' into feat/page-reports-v1.1
